app/models/game.rb

```diff
@@ -30,27 +39,44 @@ def populate_game!
     save
   end
 
-  def check?
-    white_king = pieces.find_by(type: 'King', color: 'white')
-    black_pieces = pieces.where(color: 'black')
+  # Keep track of current king.  We don't care what the other king is doing.
+  def check?(color)
+    king = pieces.find_by(type: 'King', color: color)
+    opposing_pieces = pieces.where(color: king.opposite_color, captured: false)
 
-    black_pieces.each { |black_piece| return true if black_piece.valid_move?(white_king.x_coordinate, white_king.y_coordinate) }
+    opposing_pieces.each { |opposing_piece| return true if opposing_piece.valid_move?(king.x_coordinate, king.y_coordinate) }
+    false
+  end
 
-    black_king = pieces.find_by(type: 'King', color: 'black')
-    white_pieces = pieces.where(color: 'white')
+  def checkmate?(color)
+    #  a. king has to be in check
+    return false unless check?(color)
 
-    white_pieces.each { |white_piece| return true if white_piece.valid_move?(black_king.x_coordinate, black_king.y_coordinate) }
+    #  a1. king cannot get out of check (has no more valid moves to get out of check)
+    #  a2. king cannot be blocked by another piece (any color)
+    #  a3. pieces checking king check cannot be captured
 
-    false
+    # Go through every possible move for every single piece I have and see if you can get out of checkmate
+    pieces.find_by(color: color, captured: false).each do |piece|
+      piece.valid_moves(x, y).each do |move|
+        return false unless misstep?(piece, move.x, move.y)
+      end
+    end
+    true
   end
 
-  def inactive_player
-    return player_1_id if turn == player_2_id
-    return player_2_id if turn == player_1_id
-  end
+  # Make the move and check to see if it was a mis step, and move back.
+  def misstep?(moving_piece, move_to_x, move_to_y)
+    # starting position
+    x = moving_piece.x_coordinate
+    y = moving_piece.y_coordinate
+    # move to see if it puts you in check. if true, it's a misstep
+    moving_piece.move_to!(move_to_x, move_to_y)
+    misstep = check?(moving_piece.color)
 
-  def not_your_turn
-    !inactive_player
+    # move it back to starting position
+    moving_piece.move_to!(x, y)
+    misstep
```

Comment:
If I don't move the pieces back, then I accidentally move pieces everywhere when checking for checkmate.